sinkhole-query/src/lib.rs

```diff
@@ -7,10 +7,16 @@ pub use sinkhole_core::{PrivateKey, PublicKey};
 ///
 /// Implementations of this trait simply use different
 /// cryptography libraries for various trade-offs.
-pub trait Query {
+pub trait Builder {
     /// Constructs an encrypted query.
-    fn construct(content: &[u8], public_key: sinkhole_core::PublicKey) -> &[u8];
+    fn construct(content: &[u8], public_key: sinkhole_core::PublicKey) -> Query;
 
     /// Decrypts the result returned by the PIR server.
     fn decrypt(result: &[u8], private_key: sinkhole_core::PrivateKey) -> &[u8];
 }
+
+/// This struct represents PIR queries.
+pub struct Query {
+    /// The encoded field queries.
+    fields: [Vec<u8>],
```

Comment:
Looks good. Alternatively, we can have a `Vec<Vec<u8>>`, but not a big deal since the query is meant to be serialized and passed along to the server.

Reply:
I did `[]` because we don't want the array to be tampered with. Is my thinking here wrong?

Reply:
The `fields` will be non-mutable by definition (Rust things :) ) if that's what you mean. In order to change the values of a struct field, you'd need to wrap it in a `Box` or `RefCell` - similar to what we have in the Storage:

```
 struct Storage {                                                                        
  secret_key: SecretKey,                                                              
  size: usize,                                                                        
  store: RefCell<Vec<Scalar>>,                                                        
}    
```

If `store` was defined as ` store: Vec<Scalar>,`, the contents of the vector could not be changed.